serverless/log-email-sns/src/query.ts

```diff
@@ -0,0 +1,145 @@
+
+export const init = async () => {
+  if (sequelize === null) {
+    sequelize = await sequelizeLoader()
+  }
+}
+
+/**
+ * Adds email to blacklist table if it does not exist
+ * @param recipientEmail
+ */
+export const addToBlacklist = (recipientEmail: string) => {
+  console.log(`Updating blacklist table with ${recipientEmail}`)
+  return sequelize?.query(
+    `INSERT INTO email_blacklist (recipient, created_at, updated_at) VALUES (:recipientEmail, clock_timestamp(), clock_timestamp()) 
+      ON CONFLICT DO NOTHING;`,
+    {
+      replacements: { recipientEmail },
+      type: QueryTypes.INSERT,
+    }
+  )
+}
+
+/**
+ * Updates email_messages with an error and error code
+ *
+ */
+export const updateMessageWithError = async (
+  opts: UpdateMessageWithErrorCode
+): Promise<number | undefined> => {
+  const { errorCode, timestamp, id } = opts
+
+  console.log(`Updating email_messages table. ${JSON.stringify(opts)}`)
+  const [result] =
+    (await sequelize?.query(
+      `UPDATE email_messages SET error_code=:errorCode, received_at=:timestamp, status='INVALID_RECIPIENT', updated_at = clock_timestamp()
+      WHERE id=:id 
+      AND (received_at IS NULL OR received_at < :timestamp)
+      RETURNING campaign_id;`,
+      {
+        replacements: { errorCode, timestamp, id },
+        type: QueryTypes.UPDATE,
+      }
+    )) || []
+  const campaign: { campaign_id: number } = (result || [])[0]
+  return campaign?.campaign_id
+}
+
+/**
+ *  Updates the email_messages table for successful delivery of an email.
+ *  Delivery: Amazon SES successfully delivered the email to the recipient's mail server.
+ */
+export const updateMessageWithSuccess = async (
+  metadata: Metadata
+): Promise<number | undefined> => {
+  console.log(`Updating email_messages table. ${JSON.stringify(metadata)}`)
+  // Since notifications for the same messageId can be interleaved, we only update that message if this notification is newer than the previous.
+  const [result] =
+    (await sequelize?.query(
+      `UPDATE email_messages SET received_at=:timestamp, updated_at = clock_timestamp(), status='SUCCESS' 
+      WHERE id=:id 
+      AND (received_at IS NULL OR received_at < :timestamp)
+      RETURNING campaign_id;`,
+      {
+        replacements: { timestamp: metadata.timestamp, id: metadata.id },
+        type: QueryTypes.UPDATE,
+      }
+    )) || []
+
+  const campaign: { campaign_id: number } = (result || [])[0]
+  return campaign?.campaign_id
+}
+
+export const haltCampaignIfThresholdExceeded = async (campaignId?: number) => {
+  if (campaignId === undefined) {
+    return
+  }
+  // Compute threshold
+  const [result] =
+    (await sequelize?.query(
+      `SELECT SUM(CASE WHEN status='INVALID_RECIPIENT' THEN 1 ELSE 0 END) AS invalid, COUNT(1) AS running_total FROM email_messages WHERE campaign_id=:campaignId AND status IS NOT NULL`,
+      {
```

Comment:
Consider this scenario:
- minHaltPercentage=0.1
- campaign still sending --> all status are null in email_messages
- first callback returns a bounce --> invalid count = 1, running count = 1 --> percentage Invalid = 1
- campaign get halted

Suggestion: use total count